File: cookbook/c/sha_shs.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <charm/charm.h>


int main(void)
{
    /* INPUTS */
    /* ===================================================================== */
    /* Define the path to an input text file with spherical harmonic
     * coefficients.  For details on the structure of the text file, see the
     * description of the "charm_shc_read_mtx" function in the "charm_shc"
     * module. */
    char shcs_in_file[] = "../../data/input/EGM96-degree10-mtx.txt";


    /* Maximum harmonic degree of coefficients to read.  The same degree is
     * used later in the harmonic synthesis and harmonic analysis. */
    unsigned long nmax = 10;
    /* ===================================================================== */






    /* ===================================================================== */
    printf("Closed-loop experiment -- Spherical harmonic synthesis and "
           "analysis\n");
    printf("===========================\n");


    /* Initialize a "charm_shc" structure */
    charm_shc *shcs = charm_shc_calloc(nmax, 1.0, 1.0);
    if (shcs == NULL)
    {
        fprintf(stderr, "Failed to initialize the \"charm_shc\" structure.\n");
        exit(CHARM_FAILURE);
    }


    /* Initialize a "charm_err" structure */
    charm_err *err = charm_err_init();
    if (err == NULL)
    {
        fprintf(stderr, "Failed to initialize the \"charm_err\" structure.\n");
        exit(CHARM_FAILURE);
    }


    /* Read spherical harmonic coefficients from the input text file. */
    charm_shc_read_mtx(shcs_in_file, nmax, shcs, err);
    charm_err_handler(err, 1);


    /* Now let's say we do not want to use the zero-degree term.  This can be
     * achieved simply by setting the respective "C00" coefficient to zero. */
    shcs->c[0][0] = 0.0;


    /* Compute the Gauss--Legendre grid for a given "nmax" on a sphere that is
     * "1000.0" metres above the reference sphere of spherical harmonic
     * coefficients. */
    charm_point *grd_pnt = charm_crd_point_gl(nmax, shcs->r + 1000.0);
    if (grd_pnt == NULL)
    {
        fprintf(stderr, "Failed to compute the Gauss--Legendre grid.\n");
        exit(CHARM_FAILURE);
    }


    /* Allocate the memory for the synthesized signal */
    double *f = (double *)malloc(grd_pnt->nlat * grd_pnt->nlon *
                                 sizeof(double));
    if (f == NULL)
    {
        fprintf(stderr, "malloc failure.\n");
        exit(CHARM_FAILURE);
    }


    /* Perform the synthesis.  Since the Gauss--Legendre grid resides "1000.0"
     * metres above the reference sphere of the coefficients, performed is
     * solid spherical harmonic synthesis. */
    charm_shs_point(grd_pnt, shcs, nmax, f, err);
    charm_err_handler(err, 1);


    /* Print some synthesized values */
    printf("Print some synthesized values of the signal...\n");
    size_t i = 0;
    size_t j = 0;
    printf("f(%zu, %zu) = %0.16e\n", i, j, f[i * grd_pnt->nlon + j]);


    i = grd_pnt->nlat / 2;
    j = grd_pnt->nlon / 2;
    printf("f(%zu, %zu) = %0.16e\n", i, j, f[i * grd_pnt->nlon + j]);


    /* Initialize a new structure of spherical harmonic coefficients for
     * coefficients to be computed by the harmonic analysis.  Used are the same
     * scalling constants as in "shcs". */
    charm_shc *shcs2 = charm_shc_calloc(nmax, shcs->mu, shcs->r);
    if (shcs2 == NULL)
    {
        fprintf(stderr, "Failed to initialize the \"charm_shc\" structure.\n");
        exit(CHARM_FAILURE);
    }


    /* Now use the synthesized signal and compute back its spherical harmonic
     * coefficients by harmonic analysis.  The output coefficients in "shcs2"
     * should be the same as the input ones "shcs".  Note that the signal "f"
     * resides on a sphere that is "1000.0" metres above the reference sphere.
     * Therefore, the coefficients are automatically properly rescaled to the
     * desired sphere defined by "shcs2->r". */
    charm_sha_point(grd_pnt, f, nmax, shcs2, err);
    charm_err_handler(err, 1);


    /* Now check whether "shcs" and "shcs2" are the same by computing their
     * difference degree amplitudes */
    double *dda = (double *)malloc((nmax + 1) * sizeof(double));
    if (dda == NULL)
    {
        fprintf(stderr, "malloc failure.\n");
        exit(CHARM_FAILURE);
    }
    charm_shc_dda(shcs, shcs2, nmax, dda, err);
    charm_err_handler(err, 1);


    /* Print some difference degree amplitudes */
    unsigned long n = 0;
    printf("\n\n");
    printf("Now print the difference degree amplitudes. ");
    printf("These should be very small, say, at the order of 1e-18 "
           "or less...\n");
    printf("Difference degree amplitude for harmonic degree %lu = %0.16e\n",
            n, dda[n]);
    n = 4;
    printf("Difference degree amplitude for harmonic degree %lu = %0.16e\n",
            n, dda[n]);
    n = 10;
    printf("Difference degree amplitude for harmonic degree %lu = %0.16e\n",
            n, dda[n]);


    /* We will not need some of the structures and arrays anymore, so let's
     * free them. */
    charm_crd_point_free(grd_pnt);
    charm_shc_free(shcs2);
    free(f);
    free(dda);


    printf("===========================\n\n");
    /* ===================================================================== */






    /* ===================================================================== */
    printf("\n");
    printf("Solid spherical harmonic synthesis at scattered points\n");
    printf("===========================\n");


    /* Let's create a "charm_point" structure to store 3 scattered points. */
    charm_point *sctr_pnt = charm_crd_point_calloc(CHARM_CRD_POINT_SCATTERED,
                                                   3, 3);
    if (sctr_pnt == NULL)
    {
        fprintf(stderr,
                "Failed to initialize the \"charm_point\" structure.\n");
        exit(CHARM_FAILURE);
    }


    /* Now let's feed "sctr_pnt" with some computation points (angular values
     * must be provided in radians). */
    sctr_pnt->lat[0] =  0.1;
    sctr_pnt->lat[1] =  0.436231;
    sctr_pnt->lat[2] = -0.9651;
    sctr_pnt->lon[0] =  0.0;
    sctr_pnt->lon[1] =  1.53434;
    sctr_pnt->lon[2] =  4.2316;
    sctr_pnt->r[0]   = shcs->r + 1000.0;
    sctr_pnt->r[1]   = shcs->r + 2000.0;
    sctr_pnt->r[2]   = shcs->r + 3000.0;


    /* Allocate memory to store the synthesized signal */
    f = (double *)malloc(sctr_pnt->nlat * sizeof(double));
    if (f == NULL)
    {
        fprintf(stderr, "malloc failure.\n");
        exit(CHARM_FAILURE);
    }


    /* Do the synthesis */
    charm_shs_point(sctr_pnt, shcs, nmax, f, err);
    charm_err_handler(err, 1);


    /* It's really this easy! */


    /* Print the synthesized values */
    for (size_t i = 0; i < sctr_pnt->nlat; i++)
        printf("%0.16e\n", f[i]);


    charm_crd_point_free(sctr_pnt);
    free(f);


    printf("===========================\n\n");
    /* ===================================================================== */






    /* ===================================================================== */
    printf("\n");
    printf("Solid spherical harmonic synthesis at a custom grid of points\n");
    printf("===========================\n");


    /* Initialize a charm_point structure to hold a "5 x 10" point grid. */
    grd_pnt = charm_crd_point_calloc(CHARM_CRD_POINT_GRID, 5, 10);
    if (grd_pnt == NULL)
    {
        fprintf(stderr,
                "Failed to initialize the \"charm_point\" structure.\n");
        exit(CHARM_FAILURE);
    }


    /* Define some grid points */
    for (size_t i = 0; i < grd_pnt->nlat; i++)
    {
        grd_pnt->lat[i] = M_PI_2 - ((double)i / (double)grd_pnt->nlat) * M_PI;
        grd_pnt->r[i]   = shcs->r + (double)i;
    }
    for (size_t j = 0; j < grd_pnt->nlon; j++)
    {
        grd_pnt->lon[j] = ((double)j / (double)grd_pnt->nlon) * (2.0 * M_PI);
    }


    /* Initialize an array to store the synthesized signal */
    f = (double *)malloc(grd_pnt->nlat * grd_pnt->nlon * sizeof(double));
    if (f == NULL)
    {
        fprintf(stderr, "malloc failure.\n");
        exit(CHARM_FAILURE);
    }


    /* Do the synthesis */
    charm_shs_point(grd_pnt, shcs, nmax, f, err);
    charm_err_handler(err, 1);


    /* Print the synthesized values */
    for (size_t i = 0; i < grd_pnt->nlat; i++)
        for (size_t j = 0; j < grd_pnt->nlon; j++)
            printf("%0.16e\n", f[i * grd_pnt->nlon + j]);


    charm_crd_point_free(grd_pnt);
    free(f);


    printf("===========================\n");
    /* ===================================================================== */






    /* ===================================================================== */
    printf("\n");
    printf("Solid spherical harmonic synthesis at scattered cells\n");
    printf("===========================\n");


    /* Initialize a "charm_cell" structure to hold 3 scattered cells. */
    charm_cell *sctr_cell = charm_crd_cell_calloc(CHARM_CRD_CELL_SCATTERED,
                                                  3, 3);
    if (sctr_cell == NULL)
    {
        fprintf(stderr, "Failed to initialize scattered cells.\n");
        exit(CHARM_FAILURE);
    }


    /* Define more or less randomly some scattered cells */
    /* --------------------------------------------------------------------- */
    /* The maximum latitude of the first cell */
    sctr_cell->latmax[0] =  0.323413435;

    /* The minimum latitude of the first cell */
    sctr_cell->latmin[0] =  sctr_cell->latmax[0] - 0.234323;

    /* The maximum latitude of the second cell */
    sctr_cell->latmax[1] = -0.90234320952;

    /* The minimum latitude of the second cell */
    sctr_cell->latmin[1] =  sctr_cell->latmax[1] - 0.4456;

    /* And so on... */
    sctr_cell->latmax[2] =  0.0;
    sctr_cell->latmin[2] =  sctr_cell->latmax[2] - M_PI_2;

    /* And now the same, but with the longitudes... */
    sctr_cell->lonmin[0] =  0.123456789;
    sctr_cell->lonmax[0] =  sctr_cell->lonmin[0] + 1.3235;
    sctr_cell->lonmin[1] =  4.3445234;
    sctr_cell->lonmax[1] =  sctr_cell->lonmin[1] + 0.01;
    sctr_cell->lonmin[2] =  0.0;
    sctr_cell->lonmax[2] =  sctr_cell->lonmin[2] + 2.0 * M_PI;

    /* Finally, we define spherical radii of the scattered cells.  Importantly,
     * the spherical radius is *constants* over the cells (but may vary from
     * cell to cell). */
    sctr_cell->r[0]      =  shcs->r;
    sctr_cell->r[1]      =  shcs->r + 1000.0;
    sctr_cell->r[2]      =  shcs->r + 2000.0;
    /* --------------------------------------------------------------------- */


    /* Initialize an array to store the synthesized signal */
    f = (double *)malloc(sctr_cell->nlat * sizeof(double));
    if (f == NULL)
    {
        fprintf(stderr, "malloc failure.\n");
        exit(CHARM_FAILURE);
    }


    /* Do the synthesis */
    charm_shs_cell(sctr_cell, shcs, nmax, f, err);
    charm_err_handler(err, 1);


    /* Print the synthesized values */
    for (size_t i = 0; i < sctr_cell->nlat; i++)
        printf("%0.16e\n", f[i]);


    charm_crd_cell_free(sctr_cell);
    free(f);


    printf("===========================\n\n");
    /* ===================================================================== */






    /* ===================================================================== */
    printf("\n");
    printf("Solid spherical harmonic synthesis at a grid "
           "of cells\n");
    printf("===========================\n");


    /* Initialize a "charm_cell" structure to hold a grid of cells with "15"
     * cells in the latitudinal direction and "30" cells in the longitudinal
     * direction. */
    charm_cell *grd_cell = charm_crd_cell_calloc(CHARM_CRD_CELL_GRID, 15, 30);


    /* Define some grid cells */
    for (size_t i = 0; i < grd_cell->nlat; i++)
    {
        /* The maximum cell latitudes */
        grd_cell->latmax[i] = M_PI_2 -
                              ((double)i / (double)grd_cell->nlat) * M_PI;

        /* The minimum cell latitudes */
        grd_cell->latmin[i] = M_PI_2 -
                              ((double)(i + 1) / (double)grd_cell->nlat) *
                              M_PI;

        /* The spherical radii (may vary with the latitude, but we use
         * a constant value here, because of the example that follows after
         * this one). */
        grd_cell->r[i] = shcs->r + 1000.0;
    }
    for (size_t j = 0; j < grd_cell->nlon; j++)
    {
        /* The minimum cell longitudes */
        grd_cell->lonmin[j] = ((double)j / (double)grd_cell->nlon) *
                              (2.0 * M_PI);

        /* The maximum cell longitudes */
        grd_cell->lonmax[j] = ((double)(j + 1) / (double)grd_cell->nlon) *
                              (2.0 * M_PI);
    }


    /* Initialize an array to store the synthesized signal */
    f = (double *)malloc(grd_cell->nlat * grd_cell->nlon * sizeof(double));
    if (f == NULL)
    {
        fprintf(stderr, "malloc failure.\n");
        exit(CHARM_FAILURE);
    }


    /* Do the synthesis */
    charm_shs_cell(grd_cell, shcs, nmax, f, err);
    charm_err_handler(err, 1);


    /* Print some of the synthesized values */
    /* --------------------------------------------------------------------- */
    i = 0;
    j = 10;
    printf("f(%zu, %zu) = %0.16e\n", i, j, f[i * grd_cell->nlon + j]);


    i = 4;
    j = 3;
    printf("f(%zu, %zu) = %0.16e\n", i, j, f[i * grd_cell->nlon + j]);
    /* --------------------------------------------------------------------- */


    printf("===========================\n\n");
    /* ===================================================================== */






    /* ===================================================================== */
    printf("\n");
    printf("Spherical harmonic analysis of block-mean values in cells\n");
    printf("===========================\n");


    /* In this example, we use the "grd_cell" structure and the signal "f" from
     * the previous example. */


    /* Initialize a new structure of spherical harmonic coefficients for
     * coefficients to be recovered from the harmonic analysis.  Used are the
     * same scalling constants as in "shcs". */
    shcs2 = charm_shc_calloc(nmax, shcs->mu, shcs->r);
    if (shcs2 == NULL)
    {
        fprintf(stderr, "Failed to initialize the \"charm_shc\" structure.\n");
        exit(CHARM_FAILURE);
    }


    /* Do the analysis using the method of the approximate quadrature */
    charm_sha_cell(grd_cell, f, nmax, CHARM_SHA_CELL_AQ, shcs2, err);
    charm_err_handler(err, 1);


    /* Print some of the computed coefficients.  Note that the harmonic
     * analysis with block-mean values in cells in *not* exact, hence the
     * coefficients will not be equal to the input ones. */
    /* --------------------------------------------------------------------- */
    n = 2;
    unsigned long m = 0;
    printf("C(%3lu,%3lu) = %0.16e\n", n, m, shcs2->c[m][n - m]);
    printf("S(%3lu,%3lu) = %0.16e\n", n, m, shcs2->s[m][n - m]);


    n = 9;
    m = 0;
    printf("C(%3lu,%3lu) = %0.16e\n", n, m, shcs2->c[m][n - m]);
    printf("S(%3lu,%3lu) = %0.16e\n", n, m, shcs2->s[m][n - m]);


    n = 9;
    m = 4;
    printf("C(%3lu,%3lu) = %0.16e\n", n, m, shcs2->c[m][n - m]);
    printf("S(%3lu,%3lu) = %0.16e\n", n, m, shcs2->s[m][n - m]);


    n = 9;
    m = 9;
    printf("C(%3lu,%3lu) = %0.16e\n", n, m, shcs2->c[m][n - m]);
    printf("S(%3lu,%3lu) = %0.16e\n", n, m, shcs2->s[m][n - m]);
    /* --------------------------------------------------------------------- */


    charm_crd_cell_free(grd_cell);
    charm_shc_free(shcs2);
    free(f);


    printf("===========================\n\n");
    /* ===================================================================== */






    /* Free the heap memory */
    /* ===================================================================== */
    charm_err_free(err);
    charm_shc_free(shcs);


    printf("\nGreat, all done!\n");


    exit(CHARM_SUCCESS);
    /* ===================================================================== */
}

File: cookbook/python/sha_shs.py
import numpy as np
import pyharm as ph


# INPUTS
# =============================================================================
# Define the path to an input text file with spherical harmonic coefficients.
# For details on the structure of the text file, see the documentation.
shcs_in_file = '../../data/input/EGM96-degree10-mtx.txt'


# Maximum harmonic degree of coefficients to read.  The same degree is used
# later in the harmonic synthesis and harmonic analysis.
nmax = 10
# =============================================================================


# =============================================================================
print('Closed-loop experiment -- Spherical harmonic synthesis and analysis')
print('===========================')


# Read spherical harmonic coefficients from the input text file.
shcs = ph.shc.Shc.from_file('mtx', shcs_in_file, nmax)


# Now let's say we do not want to use the zero-degree term.  This can be
# achieved simply by setting the respective "C00" coefficient to zero.
shcs.set_coeffs(n=0, m=0, c=0.0)


# Compute the Gauss--Legendre grid for a given "nmax" on a sphere that is
# "1000.0" metres above the reference sphere of spherical harmonic
# coefficients.  PyHarm represents the Gauss--Legendre grid via the
# "PointGridGL" class.  For the Driscoll--Healy grids, PyHarm has the
# "PointGridDH1" and "PointGridDH2" classes (see the documentation).
grd_pnt = ph.crd.PointGridGL(nmax, shcs.r + 1000.0)


# Perform the synthesis.  Since the Gauss--Legendre grid resides "1000.0"
# metres above the reference sphere of the coefficients, performed is
# solid spherical harmonic synthesis.
f = ph.shs.point(grd_pnt, shcs, nmax)


# Note on parallelization
# -----------------------
#
# If you compiled the library with OpenMP parallelization enabled, you can
# control the number of threads being used either by an environmental variable
# or in your Python code, for instance, like this:
#
# from threadpoolctl import threadpool_limits
# with threadpool_limits(limits=2):
#     f = ph.shs.point(grd_pnt, shcs, nmax)


# Print some synthesized values
print('Print some synthesized values of the signal...')
i = 0
j = 0
print(f'f({i}, {j}) = {f[i, j]}')


i = grd_pnt.lat.size // 2
j = grd_pnt.lat.size // 2
print(f'f({i}, {j}) = {f[i, j]}')
print('')


# Now use the synthesized signal and compute back its spherical harmonic
# coefficients by harmonic analysis.  The output coefficients in "shcs2" should
# be the same as the input ones in "shcs".  Note that the signal "f" resides on
# a sphere that is "1000.0" metres above the reference sphere "shcs.r".  The
# last input parameter therefore allows user to define the reference radius of
# the output coefficients.  In this case, this value is set to "shcs.r" to
# ensure that "shcs2.r" will be the same as "shcs.r".
shcs2 = ph.sha.point(grd_pnt, f, nmax, shcs.mu, shcs.r)


# Now check whether "shcs" and "shcs2" are the same by computing their
# difference degree amplitudes
dda = shcs.dda(shcs2)


# Print some difference degree amplitudes
n = 0
print('Now print the difference degree amplitudes.  These should be'
      'very small, say, at the order of 1e-18 or less...')
print(f'Difference degree amplitude for harmonic degree {n} = {dda[n]}')
n = 4
print(f'Difference degree amplitude for harmonic degree {n} = {dda[n]}')
n = 10
print(f'Difference degree amplitude for harmonic degree {n} = {dda[n]}')


# We will not need some of the variables anymore, so let's free the memory by
# deleting them
del grd_pnt, shcs2, f, dda


print('===========================')
# =============================================================================


# =============================================================================
print('')
print('Solid spherical harmonic synthesis at scattered points')
print('===========================')


# Create a "PointSctr" class instance to store three scattered points.  The
# instance will be created from three numpy arrays that hold the coordinates of
# the scattered points.
lat      = np.array([0.1, 0.436231, -0.9651], dtype=np.float64)
lon      = np.array([0.0, 1.53434, 4.2316], dtype=np.float64)
r        = np.array([1000.0, 2000.0, 3000.0], dtype=np.float64) + shcs.r
sctr_pnt = ph.crd.PointSctr.from_arrays(lat, lon, r)


# Do the synthesis
f = ph.shs.point(sctr_pnt, shcs, nmax)


# It's really this easy!


# Print the synthesized values
print(f'{f}')


del sctr_pnt, f
print('===========================')
# =============================================================================


# =============================================================================
print('')
print('Solid spherical harmonic synthesis at a custom grid of points')
print('===========================')

# Define some grid points as numpy arrays.  Then create a "PointGrid" class
# instance that is designed to store custom point grids.
nlat    = 5
nlon    = 10
lat     = np.linspace(np.pi / 2.0, -np.pi / 2.0, nlat)
lon     = np.linspace(0.0, 2.0 * np.pi, nlon)
r       = np.zeros((nlat,), dtype=np.float64) + shcs.r + np.arange(nlat)
grd_pnt = ph.crd.PointGrid.from_arrays(lat, lon, r)


# Do the synthesis
f = ph.shs.point(grd_pnt, shcs, nmax)


# Print the synthesized values
print(f'{f}')


del grd_pnt, f
print('===========================')
# =============================================================================


# =============================================================================
print('')
print('Solid spherical harmonic synthesis at scattered cells')
print('===========================')


# Now we define evaluation cells to synthesize area mean values.  To this end,
# we specify numpy arrays of minimum and maximum latitudes and longitudes and
# an array of spherical radii.  Using these arrays, we create a "CellSctr"
# class instance.
latmax    = np.array([0.323413435, -0.90234320952, 0.0])
latmin    = latmax - np.array([0.234323, 0.4456, np.pi / 2.0])
lonmin    = np.array([0.123456789, 4.3445234, 0.0])
lonmax    = lonmin + np.array([1.3235, 0.01, 2.0 * np.pi])
r         = np.zeros((latmin.size,), dtype=np.float64) + shcs.r + \
            np.arange(latmax.size) + 1000.0
cell_sctr = ph.crd.CellSctr.from_arrays(latmin, latmax, lonmin, lonmax, r)


# Do the synthesis
f = ph.shs.cell(cell_sctr, shcs, nmax)


# Print the synthesized values
print(f'{f}')


del cell_sctr, f
print('===========================')
# =============================================================================


# =============================================================================
print('')
print('Solid spherical harmonic synthesis at a grid of cells')
print('===========================')


# Define the grid cells and create a "CellGrid" class instance from numpy
# arrays.
nlat     = 15
nlon     = 30
latmax   = np.linspace(np.pi / 2.0, -np.pi / 2.0, nlat, False)
latmin   = np.linspace(latmax[1], -np.pi / 2.0, nlat)
lonmin   = np.linspace(0.0, 2.0 * np.pi, nlon, False)
lonmax   = np.linspace(lonmin[1], 2.0 * np.pi, nlon)
r        = np.zeros((nlat,), dtype=np.float64) + shcs.r
grd_cell = ph.crd.CellGrid.from_arrays(latmin, latmax, lonmin, lonmax, r)


# Do the synthesis
f = ph.shs.cell(grd_cell, shcs, nmax)


# Print some of the synthesized values
# -----------------------------------------------------------------------------
i = 0
j = 10
print(f'f({i}, {j}) = {f[i, j]}')


i = 4
j = 3
print(f'f({i}, {j}) = {f[i, j]}')
# -----------------------------------------------------------------------------

print('===========================')
# =============================================================================


# =============================================================================
print('')
print('Spherical harmonic analysis of block-mean values in cells')
print('===========================')


# In this example, we use the "grd_cell" instance and the signal "f" from the
# previous example


# Do the analysis using the approximate quadrature
shcs2 = ph.sha.cell(grd_cell, f, nmax, ph.sha.CELL_AQ, shcs.mu, shcs.r)


# Print some of the computed coefficients.  Note that the harmonic analysis
# with block-mean values in cells is *not* exact, hence the coefficients will
# not be equal to the input ones.
n = [2, 9, 9, 9]  # List of harmonic degrees
m = [0, 0, 4, 9]  # List of harmonic orders
cnm, snm = shcs2.get_coeffs(n, m)
for i, (n1, m1) in enumerate(zip(n, m)):
    print(f'C({n1}, {m1}) = {cnm[i]}')
    print(f'S({n1}, {m1}) = {snm[i]}')


del grd_cell, shcs2, f, shcs


print('===========================')
# =============================================================================
